Framework: Introduce "Milestone It" GitHub action

aduth opened the pull request:
Closes #9239

This pull request seeks to introduce a [GitHub Actions](https://github.com/features/actions) Docker image configuration. The intent will be that, combined with [`actions/bin/filter` merged](https://github.com/actions/bin/tree/master/filter#merged), it can be used as an action on the `pull_request` event to assign a merged pull request into the next milestone, creating the milestone if it doesn't already exist.

**Testing Instructions:**

It's a bit difficult. I tested using my own fork of the repository. Then you'll need to create a [personal access token](https://github.com/settings/tokens/new?scopes=public_repo). Then save a [Webhook payload example](https://developer.github.com/v3/activity/events/types/#webhook-payload-example-28) for a PullRequestEvent and assign the `"number"` to a real issue or pull request in the forked repository.

At which point, from the root of Gutenberg you can do:

```
GITHUB_EVENT_PATH=~/Desktop/demo-pr.json GITHUB_TOKEN=... GITHUB_REPOSITORY=aduth/gutenberg sh .github/actions/milestone-it/entrypoint.sh
```

You should expect to see that (a) a milestone for "Gutenberg 5.8" is created and (b) the issue or pull request is assigned to it.

aduth committed:
Framework: Introduce "Milestone It" GitHub action

aduth committed:
Framework: Send milestone request result to the abyss

youknowriad commented on .github/actions/milestone-it/entrypoint.sh:
```diff
@@ -0,0 +1,54 @@
+# 2. Determine next milestone.
+
+if [ minor == '9' ]; then
+	major=$((major+1))
```
we should probably set minor to zero here

aduth replied:
> we should probably set minor to zero here

Yep, good call. Fixed in dc35b0298.

youknowriad commented on .github/actions/milestone-it/entrypoint.sh:
```diff
@@ -0,0 +1,54 @@
+	-X POST \
+	-H "Authorization: token $GITHUB_TOKEN" \
+	-H "Content-Type: application/json" \
+	-d "{\"title\":\"$milestone\"}" \
```
would be good to set the description and the date automatically as well (the date might be harder)

aduth replied:
> would be good to set the description and the date automatically as well (the date might be harder)

We don't currently have a description for these milestones. Maybe something like..

>Tasks to be included in the Gutenberg 5.8 plugin release.

?

Date... would be harder 😆 I'll think on how it might be implemented, but otherwise I could imagine it to be a future enhancement.

youknowriad replied:
> Tasks to be included in the Gutenberg 5.8 plugin release.

WFM

aduth replied:
> the date automatically as well (the date might be harder)

A thought on how this might be implemented:

1. Compute the number of versions which should have elapsed since we started strictly adhering to the two week release rhythm (or, any reference version to compare against which falls in this range).
2. Use `date` command with arithmetic to offset the start date (the date of the first bi-weekly release) by the `2 * numVersionElapsed`

e.g.

```
num_versions=8 # 5.8 - 5.0
weeks=$((num_versions*2))
reference_start_date=1549411200 # 5.0 release
due=$(date -u -r $reference_start_date -v+$(echo $weeks)w)
```

```
echo $due
# Wed May 29 00:00:00 UTC 2019
```

Unsure if this will be made complicated by differences in the `date` command between OS (macOS [BSD `date`](https://www.freebsd.org/cgi/man.cgi?date) vs. Ubuntu [GNU `date`](http://man7.org/linux/man-pages/man1/date.1.html), specifically on arithmetic via [`-v`](https://stackoverflow.com/questions/498358/how-do-i-do-date-math-in-a-bash-script-on-os-x-leopard) vs. [`-d`](https://unix.stackexchange.com/questions/49053/linux-add-x-days-to-date-and-get-new-virtual-date))

youknowriad commented on .github/actions/milestone-it/entrypoint.sh:
```diff
@@ -0,0 +1,54 @@
+		-H "Authorization: token $GITHUB_TOKEN" \
+		"https://api.github.com/repos/$GITHUB_REPOSITORY/milestones" \
+		| jq ".[0] | select(.title == \"$milestone\") | .number"
+)
```
It feels like this call could be doable before the previous one to avoid the duplicate error? and the previous does return the number of the created milestone if needed? Not sure if it's better as I understand the current implementation might be simpler.

aduth replied:
> It feels like this call could be doable before the previous one to avoid the duplicate error? and the previous does return the number of the created milestone if needed? Not sure if it's better as I understand the current implementation might be simpler.

Maybe. To me, sure, it's a failed response, but it doesn't really matter, and it keeps the implementation a bit simpler.

For context, I implemented it originally as a way to avoid fetching milestones in the repository, since I knew the results could be paginated (at least when including "all" items). This was before I realized I needed the numeric ID of the milestone to be able to update the issue, at which point I added back the fetch.

aduth committed:
Framework: Reset minor to zero at next major

aduth commented:
It occurs to me: Are there pull requests we should exempt from this behavior? I'm wondering if mobile changes, for example, are considered as part of a plugin release. I suppose it's fair to say they are?

Wondering partly because there's currently an open ["Mobile v1.0" milestone](https://github.com/WordPress/gutenberg/milestone/91). Since an issue can only be part of one milestone, if we'd want to start having mobile release milestones as well, we might have to figure out how to go about distinguishing which pull request should go where.

Since that milestone appears unused, I don't think it should be an issue for the time being.

youknowriad commented:
@aduth The only things I exclude from these are the version bump commits, this is not concerning as these commits will happen in master directly with the release tool.

But I think we should exclude the PRs that are already milestone because sometimes we assign a previous milestone to do a minor, RC fixes...

aduth commented:
> But I think we should exclude the PRs that are already milestone because sometimes we assign a previous milestone to do a minor, RC fixes...

For posterity: An _elegant_ approach to this could be a separate custom action `has_milestone` which accepts as argument either a named milestone, an empty (unassigned milestone), and optionally allows wildcards. The flow could then be `pull request` -> `filter(merged)` -> `has_milestone("")` -> `milestone_it`.

Instead, for now I will probably just add in the logic to check in this script directly 😄

aduth committed:
Framework: Include milestone description text

aduth committed:
Framework: Abort milestone assignment if one is applied

aduth committed:
Framework: Ensure entrypoint is executable

aduth commented:
Pushed up a couple more changes to reflect feedback. Starting down the path to implement the `date` handling, I was first going to grumble about the new, more-difficult testing instructions, but in retrospect am glad to have gone through the effort, since it helped reveal that there would have been problems in the entrypoint not being executable (fixed in 8808812).

Once the date work is implemented, to avoid issues of OS differences in `date` implementation, it'll be best to run through the container directly, meaning it must be built as an image and run through the `docker` command.

Example (`run` command needs substitute values):

```
docker build -t milestone-it .github/actions/milestone-it
docker run -v=/path/to/pr/payload:/run -e "GITHUB_EVENT_PATH=/run/demo-pr.json" -e "GITHUB_TOKEN=..." -e "GITHUB_REPOSITORY=aduth/gutenberg" milestone-it
```

aduth committed:
Framework: Install ca-certificates to avoid cURL errors

aduth committed:
Framework: Use bash shell to avoid redirect error

https://stackoverflow.com/questions/2462317/bash-syntax-error-redirection-unexpected

aduth committed:
Framework: Compute due date for milestone

aduth commented:
I've pushed changes to assign a due date for the milestone. I think this should be effectively complete now.

aduth committed:
Framework: Remove debugging output

youknowriad reviewed (approved):
Thanks Andrew, let's try it.

aduth committed:
Framework: Search all milestone array members for title

aduth commented:
Pushed a small fix in 17b5085. The previous implementation would have relied on the first result of the milestones API request to have been the one we're searching for, and would not have worked if it was not.

As a general reflection of implementing this: It works well enough, but I feel like writing shell scripts for these sorts of tasks is cumbersome and error-prone. Part of this may be my own inexperience in crafting them, though I think it's fair to say this would be shared by future maintainers. Since Actions are simply Docker images, there's nothing stopping us from crafting the image to install Node and run JavaScript instead, where we may benefit from our collective expertise, and perhaps even tooling patterns like unit tests, etc which would otherwise not be straight-forward as proposed here. There's a negligible amount of additional overhead in increasing the number of dependencies, but I think it would be worth considering for future custom actions.
